### DISCUSSION TITLE
allowImportingTsExtensions warning

### DISCUSSION THREAD

**Discussion by dinghui-wang** [rollup/plugins#1536] [Q&A]
tsconfig.json has "noEmit" set: true,
But  warning：
(!) Plugin typescript: @rollup/plugin-typescript TS5096: Option 'allowImportingTsExtensions' can only be used when either 'noEmit' or 'emitDeclarationOnly' is set.

**Comment by davidmyersdev** [comment]
I'm also seeing this when `emitDeclarationOnly` is set to `true`.

**Comment by eveforest** [comment]
i also met this problem too.Is there any solution?
log:
```
> tsc && vite build
vite v4.4.5 building for production...
[plugin:typescript] @rollup/plugin-typescript TS5096: Option 'allowImportingTsExtensions' can only be used when either 'noEmit' or 'emitDeclarationOnly' is set.
✓ 706 modules transformed.

``` 
```json
{
  "compilerOptions": {
    "target": "ES2021",
    "useDefineForClassFields": true,
    "lib": [
      "ES2021",
      "DOM",
      "DOM.Iterable"
    ],
    "types": [
      "vite/client",
      "node"
    ],
    "module": "ESNext",
    "skipLibCheck": true,
    "declaration": true,
    /* Bundler mode */
    "moduleResolution": "bundler",
    "allowImportingTsExtensions": true,
    "resolveJsonModule": true,
    "isolatedModules": true,
    "noEmit": true,
    /* Linting */
    "strict": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "noFallthroughCasesInSwitch": true,
    "jsx": "react-jsx",
    "baseUrl": ".",
    "paths": {
      "@common/*": [
        "components/_common/*"
      ],
      "@/*": [
        "components/*"
      ],
    }
  },
  "include": [
    "components"
  ],
  "references": [
    {
      "path": "./tsconfig.node.json"
    }
  ]
}
``` 
```json
{
  "peerDependencies": {
    "antd": ">=5.6.3",
    "ramda": ">=0.29.0",
    "react": ">=18.2.0",
    "react-dom": ">=18.2.0",
    "classnames": ">=2.3.2",
    "@ant-design/icons": ">=5.0.1",
    "react-intl": ">=6.4.4",
    "uuid": ">=9.0.0",
    "ahooks": ">=3.7.6",
    "react-draggable": ">=4.4.5",
    "@dnd-kit/core": ">=6.0.8",
    "@dnd-kit/modifiers": ">=6.0.1",
    "@dnd-kit/sortable": ">=7.0.2",
    "@dnd-kit/utilities": ">=3.2.1",
    "@iconify/react": ">=4.1.1",
    "moment": ">=2.29.4",
    "react-evefyou-hooks": ">=1.0.7"
  },
  "devDependencies": {
    "@types/uuid": "^9.0.1",
    "@rollup/plugin-typescript": "^11.1.2",
    "@types/node": "^20.4.9",
    "@types/ramda": "^0.29.0",
    "@types/react": "^18.2.15",
    "@types/react-dom": "^18.2.7",
    "@typescript-eslint/eslint-plugin": "^6.0.0",
    "@typescript-eslint/parser": "^6.0.0",
    "@vitejs/plugin-react": "^4.0.3",
    "eslint": "^8.45.0",
    "less": "^4.1.3",
    "eslint-plugin-react-hooks": "^4.6.0",
    "eslint-plugin-react-refresh": "^0.4.3",
    "tslib": "^2.6.1",
    "typescript": "^5.1.6",
    "vite": "^4.4.5",
    "vite-plugin-dts": "^3.5.1",
    "vite-tsconfig-paths": "^4.2.0"
  }
}

```

**Comment by CH-PatrickWalmsley** [comment]
+1 on this issue.
My tsconfig is correct and uses allowImportingTsExtensions with noEmit.
tsc -p . parses my typescript with extensions just fine.
rollup -c throws this warning.

**Comment by littlefatcat163** [comment]
+1 on this issue.

**Comment by TrickyPi** [comment]
Setting the `noForceEmit` option to `true` can solve this problem, see https://www.npmjs.com/package/@rollup/plugin-typescript#noforceemit.

**Reply by Gianarci** [reply]
With this flag the warning disappears but the index.d.ts file isn't emitted anymore and this is a problem for me.

**Comment by vasilecampeanu** [comment]
+1

**Comment by Jeggery** [comment]
Has it been resolved？

**Comment by shark-o** [comment]
+1

**Comment by tofsjonas** [comment]
+1

**Comment by Kiyozz** [comment]
+1

**Comment by berkerdemirer** [comment]
+1

**Comment by pbstar** [comment]
+1

**Comment by thep0y** [comment]
+1

I think just a discussion won't be given much importance, it should be turned into a question to be taken seriously.

**Comment by gakki0528** [comment]
+1

**Comment by codewithbersh** [comment]
+1

**Comment by darryl-mtrs** [comment]
+1
I stuck on this issue and found this discussion which turns out not the original library discussion.
I believe this issue is related with typescript configuration? since I encounter this error when implementing integration test on my project.

Using both noEmit and allowImportingTsExtensions (even when only one of them exists) in ts config file gives me an error: `Option 'allowImportingTsExtensions' can only be used when either 'noEmit' or 'emitDeclarationOnly' is set.`

**Comment by ShotMeow** [comment]
+1

**Comment by johanmha** [comment]
+1

**Comment by shavonn** [comment]
+1

**Comment by michealroberts** [comment]
+1

**Comment by Trung-Pham-Modec** [comment]
+1

**Comment by br14n-sol** [comment]
+1

**Comment by disputenet** [comment]
This issue was first reported back in July 2023 and not resolved. Is this project still alive?

**Comment by larsbs** [comment]
This is fixed by setting `noForceEmit: true` in the options of the plugin:

```
typescript({
  noForceEmit: true,
  exclude: ['**/__tests__', '**/*.test.ts', '**/*.test.tsx', '**/*.stories.tsx'],
}),
```

It's buried within the docs: https://github.com/rollup/plugins/tree/master/packages/typescript/#noforceemit

**Comment by disputenet** [comment]
No, it isn't resolved. Please see the previous comments about setting `noForceEmit` option.

**Comment by npquynhdo** [comment]
2025, still having this problem

**Comment by shavonn** [comment]
I tried the noForceEmit. Nothing changed, though.

**Comment by oguru** [comment]
This might help people using the plugin with Vite:
If you add vite-plugins-dts: `dts({rollupTypes: true})` and the argument `noForceEmit: true` in the typescript plugin i.e. `typescript({
            noForceEmit: true
        })`,
it builds with the declaration files and doesn't show the warning

**Comment by darcyrush** [comment]
I removed this warning on two different dual package libraries by also adding the following;
```json
"rewriteRelativeImportExtensions": true
```
To summarize the [TS handbook ](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-5-7.html#path-rewriting-for-relative-paths); The `allowImportingTsExtensions` setting allows tooling (node et all) to read the relative .ts import paths when reading .ts files directly (without transpiling), but `rewriteRelativeImportExtensions` allows the tsc to continue to output the .ts import paths as .js when transpiling to .js files. At least that is my understanding. 

So my final tsconfig.json was the following; 
```json
"compilerOptions": {
    "allowImportingTsExtensions": true,
    "noEmit": true,
    "rewriteRelativeImportExtensions": true
}
```

My usecase may be different, but it neatly solved my issues.
